src/libraries/System.Formats.Tar/src/System/Formats/Tar/SubReadStream.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Formats.Tar
+{
+    // Stream that allows wrapping a super stream and specify the lower and upper limits that can be read from it.
+    // It is meant to be used when the super stream is unseekable.
+    // Does not support writing.
+    internal class SubReadStream : Stream
+    {
+        protected bool _hasReachedEnd;
+        protected readonly long _startInSuperStream;
+        protected long _positionInSuperStream;
+        protected readonly long _endInSuperStream;
+        protected readonly Stream _superStream;
+        protected bool _isDisposed;
+
+        public SubReadStream(Stream superStream, long startPosition, long maxLength)
+        {
+            if (!superStream.CanRead)
+            {
+                throw new InvalidOperationException(SR.IO_NotSupported_UnreadableStream);
+            }
```

Comment:
In what situations will this exception be thrown? Should this be an assert instead?

Reply:
Good point. We shouldn't reach all the way here with an unreadable stream. We already checked that in the public APIs.